src/v/datalake/translation/scheduling_policies.cc

```diff
@@ -63,4 +63,258 @@ ss::future<> simple_fcfs_scheduling_policy::on_resource_exhaustion(
+                   .weight = -total_running_time.count()});
+            } else {
+                // todo: perhaps we can order by pending_bytes_to_translate
+                candidates.push_back(
+                  {.id = id,
+                   .group = translator_group::other,
+                   .weight = random_generators::get_int<long>()});
+            }
+        }
+        const auto& back = candidates.back();
+        vlog(
+          datalake_log.trace,
+          "scheduling candidate: {},  group: {}, weight: {}",
+          translator,
+          back.group,
+          back.weight);
+    }
+
+    co_await ss::maybe_yield();
+
+    // Reactor friendly sort using the comparator
+    std::priority_queue<entry, chunked_vector<entry>, decltype(max_heap_cmp)>
+      prioritized(max_heap_cmp);
+    for (auto& entry : candidates) {
+        executor.as.check();
+        auto it = executor.translators.find(entry.id);
+        if (
+          it == executor.translators.end()
+          || !it->second._waiting_hook.is_linked()) {
+            continue;
+        }
+        prioritized.push(entry);
+        co_await ss::maybe_yield();
+    }
+
+    executor.as.check();
+
+    while (!prioritized.empty()) {
+        auto& entry = prioritized.top();
+        auto it = executor.translators.find(entry.id);
+        if (
+          it != executor.translators.end()
+          && it->second._waiting_hook.is_linked()) {
+            vlog(
+              datalake_log.trace,
+              "[{}] chosen translator to run, group: {}, weight: {}",
+              it->second,
+              entry.group,
+              entry.weight);
+            executor.start_translation(it->second, _translation_time_quota);
+            co_return;
+        }
+        prioritized.pop();
+        co_await ss::maybe_yield();
+    }
+}
+
+ss::future<> fair_scheduling_policy::on_resource_exhaustion(
+  executor& executor, const reservations_tracker& mem_tracker) {
+    if (!mem_tracker.memory_exhausted() || executor.running.empty()) {
+        co_return;
+    }
+    // stop the translator with highest memory usage first.
+    // The sort is stable so it is naturally ordered based on enqueue time
+    // incase of ties. note: the size of this list is super small (low single
+    // digits)
+    executor.running.sort(
+      [](const translator_executable& a, const translator_executable& b) {
+          return a.status().memory_bytes_reserved
+                 > b.status().memory_bytes_reserved;
+      });
+
+    // pick the earliest scheduled translator and force a flush.
+    if (!executor.running.empty()) {
+        vlog(
+          datalake_log.debug,
+          "[{}] stopping translator due to memory exhaustion",
+          *executor.running.begin());
+        executor.stop_translation(*executor.running.begin());
+    }
+    while (mem_tracker.memory_exhausted() && !executor.as.abort_requested()) {
+        co_await ss::sleep_abortable(polling_interval, executor.as);
+    }
```

Comment:
Are we guaranteed to make progress within the policy here? I'm wondering does something like this end up happening?

1. we have 10/10 blocks reserved
2. t_1 attempts to reserve a block, which notifies the scheduler about memory exhaustion and waits
3. since the scheduler and policy runs in a single loop, we eventually get to this line after asynchronously stopping t_big which has one block
4. t_1 immediately takes the reservation since it's been waiting on the semaphore
5. this loop doesn't exit because memory is still exhausted now that t_1 has taken the reservation, until the translators naturally finish
6. ?? maybe other translators try to reserve, but the scheduler is stuck in this loop, so no translator can make progress?

Reply:
Right this case is possible I think if all the translators are really high throughput. As for progress,  the wait time is bounded here (meaning we are not stuck in this loop forever) because each translator will naturally release memory as its time slice finishes. I have an idea to improve the behavior here, let me push in next rev.

Reply:
Hmm I think I'm missing something. What bounds the wait time? It doesn't look like this while loop ever exits if we get to step 6 (it'd be a live lock), given neither the `default_reservation_tracker::reserve_memory()` nor this while loop have a timeout?

Reply:
The loop exits because `mem_tracker.memory_exhausted()` returns false  eventually when the inflight translations exceed their time quota and release their resources (see mock_translator impl).

Reply:
Could we get into a scenario where all translators are waiting on the reservation semaphore with no deadline, while this loop is also waiting for memory to be freed with no deadline?

Reply:
I don't think thats possible because there is a deadline which is the time_slice allotted to every translator and the translator has to release its resources after that time slice elapses no matter what which keeps the wait here bounded.

Reply:
Ahh I think you're right. I missed that the abort source used by the reservation tracker is the translator abort source, not the executor's. Sorry for the noise!